### PR TITLE
Feature doc schemes

### PR DESCRIPTION
**Pull request by lukasdreyer**
**_Describe your changes here:_**
Updated the doxygen comments for `src/t8_element_cxx.hxx` and used these as a basis to have doxygen comments for each of the different eclass schemes

**_All these boxes must be checked by the reviewers before merging the pull request:_**

- [x] The author added a BSD statement to `doc/` (or already has one)
- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request indroduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The reviewer executed the new code features at least once and checked the results manually
- [ ] The code is covered in an existing or new test case
- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [ ] Testing of this template: If you feel something is missing from this list, contact the developers

### CODE REVIEW

**Comment by lukasdreyer** [comment]
TODO: Remove SC_ABORT from functions that are implemented in `t8_default_common_cxx.cxx`

File: src/t8_schemes/t8_default/t8_default_hex/t8_default_hex_cxx.hxx [hunk 1]
```diff
@@ -216,54 +437,123 @@ public:
   virtual int         t8_element_is_root_boundary (const t8_element_t *elem,
                                                    int face);
 
-  /** Construct the face neighbor of a given element if this face neighbor
-   * is inside the root tree. Return 0 otherwise. */
+    /** Construct the face neighbor of a given element if this face neighbor
```
Comment (Davknapp) [inline src/t8_schemes/t8_default/t8_default_hex/t8_default_hex_cxx.hxx:440]: This is not properly indented

File: src/t8_schemes/t8_default/t8_default_prism/t8_default_prism_cxx.hxx [hunk 2]
```diff
@@ -46,262 +46,563 @@ public:
...
+   */
   virtual void        t8_element_copy (const t8_element_t *source,
                                        t8_element_t *dest);
 
-/** Compare to elements. returns negativ if elem1 < elem2, zero if elem1 equals elem2
- *  and positiv if elem1 > elem2.
- *  If elem2 is a copy of elem1 then the elements are equal.
- */
+  /** Compare two elements.
+    * \param [in] elem1  The first element.
```
Comment (Davknapp) [inline src/t8_schemes/t8_default/t8_default_prism/t8_default_prism_cxx.hxx:110]: This is not properly indented

File: src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx [hunk 3]
```diff
@@ -44,293 +44,573 @@ public:
...
+  /** Compute the number of siblings of an element. That is the number of 
+   * Children of its parent.
+   * \param [in] elem The element.
+   * \return          The number of siblings of \a element.
+   * Note that this number is >= 1, since we count the element itself as a sibling.
+   */
+  virtual int         t8_element_num_siblings (const t8_element_t *elem)
+    const;
+
+  /** Compute the number of siblings of an element. That is the number of
```
Comment (Davknapp) [inline src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx:122]:
```suggestion
  /** Compute a specific sibling of \a elem
```

File: src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx [hunk 4]
```diff
@@ -44,293 +44,573 @@ public:
...
+   * Children of its parent.
+   * \param [in] elem The element.
+   * \return          The number of siblings of \a element.
+   * Note that this number is >= 1, since we count the element itself as a sibling.
+   */
+  virtual int         t8_element_num_siblings (const t8_element_t *elem)
+    const;
+
+  /** Compute the number of siblings of an element. That is the number of 
+   * Children of its parent.
```
Comment (Davknapp) [inline src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx:123]:
```suggestion
```

File: src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx [hunk 5]
```diff
@@ -44,293 +44,573 @@ public:
...
+   * \param [in] elem The element.
```
Comment (Davknapp) [inline src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx:124]:
```suggestion
   * \param [in] elem Input element
   * \param [in] sibid The id of the sibling computed, 0...9 in Bey order
   * \param [in, out] Existing pyramid whose data will be filled with the data of sibling no \a sibid of \a elem
```

Comment (Davknapp) [reply src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx:124]: This documentation needs to be updated for every element-class

**Review by Davknapp — COMMENTED** [review]
Looks good in general! Most of my commenty apply to each element-file as the documentation seems to be copy pasted. If this is fixed, it should be ready to merge

File: src/t8_schemes/t8_default/t8_default_hex/t8_default_hex_cxx.hxx [hunk 6]
```diff
@@ -93,117 +153,278 @@ public:
...
+   * \param [in] boundary_scheme The scheme for the eclass of the boundary face.
+   */
   virtual void        t8_element_boundary_face (const t8_element_t *elem,
                                                 int face,
                                                 t8_element_t *boundary,
                                                 const t8_eclass_scheme_c
                                                 *boundary_scheme);
 
-/** Construct all codimension-one boundary elements of a given element. */
+  /** Construct all codimension-one boundary elements of a given element. */
```
[the remark this answers is not in the document]
Comment (lukasdreyer) [reply src/t8_schemes/t8_default/t8_default_hex/t8_default_hex_cxx.hxx:427]: This function has the following comment in `t8_element_cxx.hxx`:
```
  /* TODO:  Do we need this function at all?
   *        If not remove it. If so, what to do with prisms and pyramids?
   *        Here the boundary elements are of different eclasses, so we cannot
   *        store them in an array...
   */
```
Do you know if we already have an Issue to cleanup the element functions or should I create a new Issue?

File: src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx (hunk 3 above)
Comment (lukasdreyer) [reply src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx:122]: I am not sure that I understand this suggestion

File: src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx (hunk 4 above)
Comment (lukasdreyer) [reply src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx:123]: I am not sure that I understand this suggestion

File: src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx (hunk 3 above)
Comment (lukasdreyer) [reply src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx:122]: I have understood it now

File: src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx (hunk 4 above)
Comment (lukasdreyer) [reply src/t8_schemes/t8_default/t8_default_pyramid/t8_default_pyramid_cxx.hxx:123]: I have understood it now

File: src/t8_schemes/t8_default/t8_default_hex/t8_default_hex_cxx.hxx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_schemes/t8_default/t8_default_hex/t8_default_hex_cxx.hxx:427]: I dont know about such Issue, we should create one!

**Review by Davknapp — COMMENTED** [review]
Looks good so far, we can merge after adressing the last comment and after I have checked it on my computer

File: src/t8_schemes/t8_default/t8_default_vertex/t8_default_vertex_cxx.hxx [hunk 7]
```diff
@@ -88,165 +152,324 @@ public:
...
+  /** Return the corner number of an element's face corner.
+   * \param [in] element  The element.
+   * \param [in] face     A face index for \a element.
+   * \param [in] corner   A corner index for the face 0 <= \a corner < num_face_corners.
+   * \return              The corner number of the \a corner-th vertex of \a face.
+   */
   virtual int         t8_element_get_face_corner (const t8_element_t *element,
                                                   int face, int corner)
   {
-    SC_ABORT_NOT_REACHED ();    /* it is impossible to have a face of a vertex */
```
[the remark this answers is not in the document]
Comment (lukasdreyer) [reply src/t8_schemes/t8_default/t8_default_vertex/t8_default_vertex_cxx.hxx:101]: I concur! This resulted from copy-pasting not only the documentation but the whole file and fixing the missing functionalities afterwards, but I did not see that there were different possibilities to abort.

**Review by Davknapp — APPROVED** [review]

**Pull request merged**